drned/common/test_template.py

```diff
@@ -242,39 +241,59 @@ def _drned_single_set(device, init, fname, init_op, op, end_op, it, ordered):
         set_all_operation = True
         # Use upper scheme as default
         the_end = end_op
-        for tset_file in tset_files:
-            src,eop,sop = _drned_single_file(device, tset_file, op)
-            src_in_set = src_in_set or src
-            set_all_operation = set_all_operation and sop
-            # Get end operations from file, if present
-            if eop:
-                # File operations override
-                the_end = eop
-        if the_end == None:
-            # Default action is to first rollback to start, then to
-            # end, then rollback all commits, one by one
-            if set_all_operation:
-                the_end = ["rollback+0", "commit", "compare-config"]
-                extra_rollback = 1
+        try:
+            for tset_file in tset_files:
+                src,eop,sop = _drned_single_file(device, tset_file, op)
+                src_in_set = src_in_set or src
+                set_all_operation = set_all_operation and sop
+                # Get end operations from file, if present
+                if eop:
+                    # File operations override
+                    the_end = eop
+
+            if the_end == None:
+                # Default action is to first rollback to start, then to
+                # end, then rollback all commits, one by one
+                if set_all_operation:
+                    the_end = ["rollback+0", "commit", "compare-config"]
+                    extra_rollback = 1
+                else:
+                    the_end = []
+                    extra_rollback = 0
+                for i in reversed(range(commit_id_base, len(device.commit_id) + extra_rollback)):
+                    the_end += ["rollback+%d" % (i - commit_id_base), "commit", "compare-config"]
+            # Use the file function also for rollbacks
+            _drned_single_file(device, None, the_end,
+                               commit_id_base=commit_id_base)
+        except BaseException:
+            # Record the failed state and, if this is not the last
+            # state in the list, restore the device to its original
+            # configuration and continue testing. If this is the last
+            # test, re-raise the exception and trigger regular
+            # cleanup/error reporting.
+            device.failed_states.append(os.path.basename(tset_file))
```

Comment:
This means that the user does not know about the failure until the very end; isn't it possible to print a failure message and process it in the filtering part?

Reply:
Well, it does print an error message and the cleanup processing log, at least when running at the drned-overview log-level. I'll have a look and see what I can do to make it more clear.